confd/service/service.go

```diff
@@ -0,0 +1,188 @@
+	return &Service{
+		Name: raw["name"].(string),
+		URL:  "http://" + raw["service"].(string),
+	}
+}
+
+func convertToServices(kapi client.KeysAPI, tag string, key string) (Services, error) {
+	resp, err := kapi.Get(context.Background(), key, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read service key %s from etcd", key)
+	}
+
+	services := Services{}
+	for _, child := range resp.Node.Nodes {
+		service, err := convertToService(tag, child.Value)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert node to service")
+		} else if service != nil {
+			services = append(services, service)
+		}
+	}
+
+	return services, nil
+}
+
+func convertToService(tag string, value string) (*Service, error) {
+	raw := RawData{}
+	err := json.Unmarshal([]byte(value), &raw)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall service json")
+	}
+
+	if tag != "" {
+		if raw["tags"] != nil {
+			tags := raw["tags"].([]interface{})
+			if Contains(tags, tag) {
+				return createService(raw), err
+			}
+		} else {
+			return createService(raw), err
+		}
+	}
+	return nil, nil
+}
+
+// serviceReader reads from etcd and converts the keys and value to service
+// struct, which can easily used for configuration templates
+func serviceReader(source Source, tag string, kapi client.KeysAPI) (interface{}, error) {
+	resp, err := kapi.Get(context.Background(), source.Path, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read root %s from etcd", source.Path)
+	}
+
+	services := Services{}
+	for _, child := range resp.Node.Nodes {
+		serviceEntries, err := convertToServices(kapi, tag, child.Key)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to convert node to service")
+		}
+		for _, service := range serviceEntries {
+			services = append(services, service)
+		}
+	}
+
+	return services, nil
+}
+
+func readFromConfig(configuration Configuration, kapi client.KeysAPI) (interface{}, error) {
+	return serviceReader(configuration.Source, configuration.Tag, kapi)
+}
+
+// templateWriter transform the data with a golang template
+func templateWriter(conf Configuration, data interface{}) error {
+	if conf.PreCommand != "" {
+		err := preCheck(conf, data)
+		if err != nil {
+			return errors.Wrap(err, "pre check failed")
+		}
+	}
+
+	err := write(conf, data)
+	if err != nil {
+		return errors.Wrap(err, "failed to write data")
+	}
+
+	if conf.PostCommand != "" {
+		err = post(conf.PostCommand)
+		if err != nil {
+			return errors.Wrap(err, "post command failed")
+		}
+	}
+	return nil
+}
+
+func write(conf Configuration, data interface{}) error {
+	name := path.Base(conf.Template)
+	tmpl, err := template.New(name).ParseFiles(conf.Template)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse template")
+	}
+
+	file, err := os.Create(conf.Target)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create target file %s", conf.Target)
+	}
+	defer file.Close()
```

Comment:
**[checkstyle]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
error return value not checked (defer file.Close())